parachain/pallets/control/src/lib.rs

```diff
@@ -299,6 +356,27 @@ pub mod pallet {
 
 			Ok(())
 		}
+
+		/// - `update`: (ControlOperation, Fee).
+		#[pallet::call_index(6)]
+		#[pallet::weight(T::WeightInfo::update_operation_fee())]
+		pub fn update_operation_fee(
```

Comment:
I think it is simpler to have a reconfigurable _base fee_, and then a static per-operation fee multiplier. This makes the updating the fees simpler, as there is only one variable to adjust.

For example:

```
BaseFee=10 DOT
Fee(CreateAgent) = BaseFee * StaticMultiplier[CreateAgent]
Fee(UpdateChannel) = BaseFee * StaticMultiplier[UpdateChannel]
Fee(TransferToken) = BaseFee * StaticMultiplier[TransferToken]
```

@claravanstaden is also working on fees charged by the outbound router for the `TransferToken` operation, and so I think both your work needs to align on the implementation somehow.

Assuming we go for the BaseFee approach, we also need to store the base fee somewhere. I think the `outbound-router` pallet is the best place. The pallet could implement a trait like the following, so that other pallets can access the base fee.

```rust
trait FeeProvider {
  fn base_fee() -> Balance;
}
```

Reply:
That makes sense! I'll handle that and make it reusable for other outbound operations like transfer.

Reply:
I think it's also important to define `BaseFee` as something we can reason about, for example:

_Cost in DOT for 500,000 GAS_

Reply:
So from the previous comment, even BaseFee is composed of several factors:

* DOT/ETH price
* Gas Price (Ether)
* Gas Amount

It may make sense to thus define BaseFee in terms of those three sub-parameters.

Reply:
Yes, the DOT/ETH price and Gas Price (Ether) are some common parameters that apply to all operations,  and the gas amount for each operation can be extracted from the gas report.

Btw, the `outbound-router` currently in our repo is some primitives and not a pallet. So I would like to add another pallet for that.  On-chain storage in that pallet including:

- a dynamic map from operation type to gas amount(e.g. CreateAgent->500 gas, TransferToken->100 gas,...)
- a configurable price ratio(reflect the two price above).

I would assume the trait should be modified to something like

```
trait FeeProvider {
  fn base_fee(operation Operation) -> Balance;
}
```
and the default implemenation in the pallet for `base_fee` is simply to be `price_ratio` multiply `gas_amount` for that operation.

@vgeddes Please let me know what you think.

Reply:
> Btw, the outbound-router currently in our repo is some primitives and not a pallet. So I would like to add another pallet for that.

Oops, I meant the `outbound-queue` pallet. This is the right place for it.

Reply:
> a dynamic map from operation type to gas amount(e.g. CreateAgent->500 gas, TransferToken->100 gas,...)

I don't want to introduce this level complexity just yet, as we need to have this PR ready by Wednesday at the latest.

We just need an approximate solution that satisfies the audit review.

So the `control` pallet computes the fee for `CreateAgent` by executing:

```
let fee = T::FeeProvider::base_fee() * CREATE_AGENT_MULTIPLIER
```

Reply:
> So from the previous comment, even BaseFee is composed of several factors:
> 
> * DOT/ETH price
> * Gas Price (Ether)
> * Gas Amount
> 
> It may make sense to thus define BaseFee in terms of those three sub-parameters.

So to be more clear, I suggest calculating a global `BaseFee` using a _static_ gas amount of "100,000".

Then if `CreateAgent` uses 200,000 gas, then `CREATE_AGENT_MULTIPLIER` is `2`.

Reply:
All of this sounds good 😄

Reply:
https://github.com/Snowfork/snowbridge/pull/940/commits/cb6f6aaa5d615a53e4f53d0ef53566c71b9ff75e